security/pkg/pki/ca/ca_test.go

```diff
@@ -283,7 +283,8 @@ func TestCreateSelfSignedIstioCAReadSigningCertOnly(t *testing.T) {
 }
 
 func TestCreatePluggedCertCA(t *testing.T) {
-	t0 := time.Now()
+	// call .UTC() to trim monotonic clock reading,
+	t0 := time.Now().UTC()
```

Comment:
I think this doesn't really fix the flaky. I patched your fix locally and still find the test failure:
```
$ go test -run TestCreatePluggedCertCA -count 3000
--- FAIL: TestCreatePluggedCertCA (0.00s)
    ca_test.go:331: Invalid default cert TTL, should be the same as cert chain: 2028-02-23 08:53:51.999860614 +0000 UTC VS (expected) 2028-02-23 08:53:52 +0000 UTC
--- FAIL: TestCreatePluggedCertCA (0.00s)
    ca_test.go:331: Invalid default cert TTL, should be the same as cert chain: 2028-02-23 08:53:51.999592619 +0000 UTC VS (expected) 2028-02-23 08:53:52 +0000 UTC
FAIL
exit status 1
FAIL	istio.io/istio/security/pkg/pki/ca	3.855s
```

---

 I think the real issue here is because the `ca.defaultCertTTL` is calculated by subtracting the `certChain.NotAfter` with a call to `time.Now()` in NewIstioCA() in https://github.com/istio/istio/blob/6cff49fa4ba6a3a78f0b6849213bfbb74e5f422b/security/pkg/pki/ca/ca.go#L383, and later the duration is computed from another `time.Now()` called in the beginning of this function in line 287 `t0: time.Now()`.

The two separate `time.Now()` call could have a gap (probably due to the `NewPluggedCertIstioCAOptions()` takes too long to return) that eventually caused the test to fail.

If you just move the `t0 := time.Now()` to be just above the `ca, err := NewIstioCA(caopts)`, it actually significantly reduces the flaky occurrences (I couldn't reproduce even with `go test -run TestCreatePluggedCertCA -count 5000`).

But that is not a very good fix as it assumes the two time.Now() happens soon enough (< 1s). The proper fix is that, in addition to move the `t0 := time.Now()` to be above the NewIstioCA, you can compare the `t0.Add(ca.defaultCertTTL).Unix()` and `certChain.NotAfter.Unix()` are within a reasonable range (e.g. < 10s), this works because we set the `defaultWorkloadCertTTL := 99999 * time.Hour` which is far smaller than the `certChain.NotAfter` (2028-02-23).

cc  @lucasvmiguel who added the test in the first place.

Reply:
Nice catch! Thanks for you detailed explanation! I have updated the code as you said.